### PR TITLE
FTUE - Server selection errors

### LINKED ISSUE #5749: Server selection - Handle disable registration
Within the new Server Selection screen, we should handle disabled registration errors within the page to avoid the user finding out after typing in a username/password

- When selecting a server during account creation
- If the server does not accept registration, show the error within the server selection input field (like we did in the legacy flow)

| LEGACY | FTUE | 
| --- | --- |
Screenshot_20220412_171436|Screenshot_20220412_171214

### PR DESCRIPTION
**Pull request by ouchadam**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

- Uses the TextUri type for url entry (fixes spaces being added when typing)
- Adds missing error states to the `Server Selection` screen instead of when submitting the account details
  - Starts the registration flow at the point of setting the homeserver url instead of waiting until details submission  
- Fixes flashing loading state when selecting a usage option by flattening nested scope launching and only applying loading changes to top level entry points

## Motivation and context

Fixes #5749

## Screenshots / GIFs

|Before|After|
|-|-|
|before-disabled-reg|after-disabled-reg|
|before-wrong-url|after-wrong-url

## Tests

- With the combined register feature flag enabled
- Start the account creation flow
- Edit the server and supply a server with registration disabled or an invalid url
- Notice the error message is now displayed within the entry field

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s): 29

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/onboarding/OnboardingViewModel.kt [hunk 1]
```diff
@@ -255,41 +255,52 @@ class OnboardingViewModel @AssistedInject constructor(
         }
     }
 
-    private fun handleRegisterAction(action: RegisterAction) {
+    private fun handleRegisterAction(action: RegisterAction, onNextRegistrationStepAction: (FlowResult) -> Unit) {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/onboarding/OnboardingViewModel.kt:258]: the callers of `handleRegisterAction` can decide how to the handle the next step (such as emitting flow responses or closing the edit selection screen)

File: vector/src/main/java/im/vector/app/features/onboarding/OnboardingViewModel.kt [hunk 2]
```diff
@@ -581,42 +591,77 @@ class OnboardingViewModel @AssistedInject constructor(
...
+    ) {
+        rememberHomeServer(config.homeServerUri.toString())
+        if (authResult.isHomeserverOutdated) {
+            _viewEvents.post(OnboardingViewEvents.OutdatedHomeserver)
+        }
+
+        when (trigger) {
+            is OnboardingAction.HomeServerChange.EditHomeServer   -> {
+                when (awaitState().onboardingFlow) {
+                    OnboardingFlow.SignUp -> internalRegisterAction(RegisterAction.StartRegistration) { _ ->
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/onboarding/OnboardingViewModel.kt:631]: adding the `internalRegisterAction` to the edit server flow is the fix for the missing errors, updating the server selection after success also ensures when closing the screen the previous homeserver selection in not overwritten

File: vector/src/test/java/im/vector/app/features/onboarding/OnboardingViewModelTest.kt [hunk 3]
```diff
@@ -235,24 +231,49 @@ class OnboardingViewModelTest {
...
+                        { copy(selectedHomeserver = SELECTED_HOMESERVER_STATE) },
+                        { copy(isLoading = false) }
+
                 )
                 .assertEvents(OnboardingViewEvents.OnHomeserverEdited)
                 .finish()
     }
 
+    @Test
+    fun `given in the sign up flow, when editing homeserver errors, then does not update the selected homeserver state and emits error`() = runTest {
```
Comment (ouchadam) [inline vector/src/test/java/im/vector/app/features/onboarding/OnboardingViewModelTest.kt:257]: test to ensure we don't set invalid homeservers when editing fails

File: vector/src/main/java/im/vector/app/features/onboarding/OnboardingViewModel.kt [hunk 4]
```diff
@@ -255,41 +256,52 @@ class OnboardingViewModel @AssistedInject constructor(
         }
     }
 
-    private fun handleRegisterAction(action: RegisterAction) {
+    private fun handleRegisterAction(action: RegisterAction, onNextRegistrationStepAction: (FlowResult) -> Unit) {
         currentJob = viewModelScope.launch {
             if (action.hasLoadingState()) {
                 setState { copy(isLoading = true) }
             }
-            runCatching { registrationActionHandler.handleRegisterAction(registrationWizard, action) }
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/onboarding/OnboardingViewModel.kt:263]: this diff in this function is a bit confusing, essentially this logic was extracted to a suspending `internalRegisterAction` which other `handle*` entry points can reuse

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt [hunk 5]
```diff
@@ -99,6 +100,11 @@ fun Throwable.isInvalidUIAAuth(): Boolean {
             error.flows != null
 }
 
+fun Throwable.isHomeserverUnavailable(): Boolean {
+    return this is Failure.NetworkConnection &&
+            this.ioException is UnknownHostException
+}
+
```
Comment (ericdecanini) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt:107]: Optional: I think the `fun Throwable.isHomeserverUnavailable() = this is...` syntax would be nicer here

**Review by ericdecanini — APPROVED** [review]
LGTM! I also like that we extracted the handle register logic and increased its usability within the class ✨

Left a very minor optional but that's really up to you

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt (hunk 5 above)
Comment (ericdecanini) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/failure/Extensions.kt]: Thanks for making this change. Looking awesome!

**Comment by ouchadam** [comment]
lint failed with 

```
* Exception is:
org.gradle.launcher.daemon.client.DaemonDisappearedException: Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)
```

will rerun :crossed_fingers:

**Pull request merged**